packages/core/src/UserState.ts

```diff
@@ -38,26 +39,32 @@ export default class UserState {
                   db?: DB
                   attesterId?: bigint | bigint[]
                   unirepAddress: string
-                  prover: Prover
                   provider: ethers.providers.Provider
-                  _id?: Identity // TODO: remove this and only accept as second arg
               }
             | Synchronizer,
-        id: Identity
+        userConfig: {
+            id: Identity
+            prover: Prover
+        }
     ) {
+        const { id, prover } = userConfig
+        if (!id) {
+            throw new Error(
+                '@unirep/core:UserState: id must be supplied as an argument when initialized'
+            )
+        }
+        if (!prover) {
```

Comment:
This constructor is now a bit of a mess. I think we should look for either `userConfig.prover` or `config.prover` to maintain backward compatibility.

Reply:
The user might supply a config as the first argument (not necessarily a synchronizer). The prover could be in this argument.

If we're going to break backward compatibility we should make this a bit more clean?

Maybe like

```ts
constructor(config: {
  synchronizer?: Synchronizer,
  db?: DB,
  attesterId?: bigint | string,
  unirepAddress?: bigint | string,
  provider?: ethers.providers.Provider,
  id: Identity,
  prover: Prover
}) {
  if (config.synchronizer) {
    this.sync = config.synchronizer
  } else {
    this.sync = new Synchronizer(config)
  }
  this.id = config.id
  this.prover = config.prover
}
```